File: platform/integrations/ios/Sources/LayerXMobile/ReceiptGate.swift
import Foundation
import LayerXSDK

public struct MobileReceiptEvidence: Sendable {
    public let canonicalReceipt: Data
    public let authorizedBatch: AuthorizedReceiptBatch

    public init(canonicalReceipt: Data, authorizedBatch: AuthorizedReceiptBatch) {
        self.canonicalReceipt = canonicalReceipt
        self.authorizedBatch = authorizedBatch
    }
}

public struct MobileSettlementExpectation: Sendable, Equatable {
    public let asset: Data
    public let recipient: Data
    public let amount: String

    public init(asset: Data, recipient: Data, amount: String) throws {
        guard asset.count == 32, recipient.count == 32,
              !amount.isEmpty, amount.utf8.allSatisfy({ $0 >= 48 && $0 <= 57 }),
              amount == "0" || amount.first != "0" else {
            throw MobileIntegrationError(.invalidConfiguration)
        }
        self.asset = asset
        self.recipient = recipient
        self.amount = amount
    }
}

public enum MobileSettlementState: Sendable, Equatable {
    case pending(reference: String)
    case verified(level: String, receiptDigest: String)
    case refused(code: String)
}

public protocol MobileReceiptResolver: Sendable {
    func resolve(receiptReference: String) async throws -> MobileReceiptEvidence
}

public struct ReceiptGate: Sendable {
    private let receipts: MobileReceiptResolver

    public init(receipts: MobileReceiptResolver) {
        self.receipts = receipts
    }

    public func settle(
        receiptReference: String,
        expecting expectation: MobileSettlementExpectation
    ) async throws -> MobileSettlementState {
        guard !receiptReference.isEmpty, receiptReference.utf8.count <= 512, !receiptReference.contains("\0") else {
            throw MobileIntegrationError(.decodeFailure)
        }
        let evidence = try await receipts.resolve(receiptReference: receiptReference)
        return try await verify(evidence: evidence, expecting: expectation)
    }

    public func verify(
        evidence: MobileReceiptEvidence,
        expecting expectation: MobileSettlementExpectation
    ) async throws -> MobileSettlementState {
        let verification: ReceiptVerification
        do {
            verification = try await LocalVerifier.verifyReceipt(evidence.canonicalReceipt, authorized: evidence.authorizedBatch)
        } catch {
            throw MobileIntegrationError(.verificationFailure)
        }
        guard verification.receipt.asset == expectation.asset,
              verification.receipt.to == expectation.recipient,
              decimalString(verification.receipt.amount) == expectation.amount else {
            throw MobileIntegrationError(.verificationFailure)
        }
        return .verified(level: verification.level, receiptDigest: hexadecimal(verification.receiptDigest))
    }

    public func project(journey: JSONValue, expecting expectation: MobileSettlementExpectation) async throws -> MobileSettlementState {
        guard let object = journey.objectValue, case let .string(state)? = object["state"] else {
            throw MobileIntegrationError(.decodeFailure)
        }
        switch state {
        case "settled", "completed":
            guard case let .string(reference)? = object["receipt_ref"] else {
                throw MobileIntegrationError(.decodeFailure)
            }
            return try await settle(receiptReference: reference, expecting: expectation)
        case "failed", "expired", "refused":
            return .refused(code: state)
        default:
            guard case let .string(reference)? = object["journey_id"] else {
                throw MobileIntegrationError(.decodeFailure)
            }
            return .pending(reference: reference)
        }
    }
}

public func decimalString(_ value: UInt128Value) -> String {
    var high = value.high
    var low = value.low
    if high == 0 && low == 0 { return "0" }
    var digits: [UInt8] = []
    while high != 0 || low != 0 {
        let quotientHigh = high / 10
        let remainderHigh = high % 10
        let (quotientLow, remainder) = UInt64(10).dividingFullWidth((high: remainderHigh, low: low))
        digits.append(UInt8(48 + remainder))
        high = quotientHigh
        low = quotientLow
    }
    return String(decoding: digits.reversed(), as: UTF8.self)
}

public func hexadecimal(_ bytes: Data) -> String {
    bytes.map { String(format: "%02x", $0) }.joined()
}

File: platform/integrations/android/src/main/java/com/sidiora/layerx/android/ReceiptGate.java
package com.sidiora.layerx.android;

import com.fasterxml.jackson.databind.JsonNode;
import com.sidiora.layerx.sdk.ProtocolAmount;
import com.sidiora.layerx.sdk.verify.LocalVerifier;
import java.math.BigInteger;
import java.security.MessageDigest;
import java.util.Objects;

/** Settlement is believed only after the device itself verifies the protocol receipt. */
public final class ReceiptGate {
    public record Evidence(byte[] canonicalReceipt, LocalVerifier.AuthorizedReceiptBatch authorizedBatch) {
        public Evidence {
            Objects.requireNonNull(canonicalReceipt, "canonicalReceipt");
            Objects.requireNonNull(authorizedBatch, "authorizedBatch");
        }
    }

    public record Expectation(byte[] asset, byte[] recipient, ProtocolAmount amount) {
        public Expectation {
            if (asset == null || asset.length != 32 || recipient == null || recipient.length != 32
                    || amount == null) {
                throw MobileIntegrationException.of(MobileIntegrationException.Code.INVALID_CONFIGURATION);
            }
            asset = asset.clone();
            recipient = recipient.clone();
        }
    }

    public sealed interface State permits Pending, Verified, Refused {}
    public record Pending(String reference) implements State {}
    public record Verified(String level, String receiptDigest) implements State {}
    public record Refused(String code) implements State {}

    @FunctionalInterface public interface ReceiptResolver {
        Evidence resolve(String receiptReference);
    }

    private final ReceiptResolver receipts;

    public ReceiptGate(ReceiptResolver receipts) {
        this.receipts = Objects.requireNonNull(receipts, "receipts");
    }

    public State settle(String receiptReference, Expectation expectation) {
        if (receiptReference == null || receiptReference.isEmpty() || receiptReference.length() > 512
                || receiptReference.indexOf('\0') >= 0) {
            throw MobileIntegrationException.of(MobileIntegrationException.Code.DECODE_FAILURE);
        }
        return verify(receipts.resolve(receiptReference), expectation);
    }

    public State verify(Evidence evidence, Expectation expectation) {
        Objects.requireNonNull(evidence, "evidence");
        Objects.requireNonNull(expectation, "expectation");
        LocalVerifier.ReceiptVerification verification;
        try {
            verification = LocalVerifier.verifyReceipt(evidence.canonicalReceipt(), evidence.authorizedBatch());
        } catch (RuntimeException error) {
            throw MobileIntegrationException.of(MobileIntegrationException.Code.VERIFICATION_FAILURE);
        }
        LocalVerifier.ProtocolReceipt receipt = verification.receipt();
        BigInteger expected = expectation.amount().value();
        if (!MessageDigest.isEqual(receipt.asset(), expectation.asset())
                || !MessageDigest.isEqual(receipt.to(), expectation.recipient())
                || !expected.equals(receipt.amount())) {
            throw MobileIntegrationException.of(MobileIntegrationException.Code.VERIFICATION_FAILURE);
        }
        return new Verified(verification.level().wire(),
            PublishableConfiguration.hexadecimal(verification.receiptDigest()));
    }

    public State project(JsonNode journey, Expectation expectation) {
        Objects.requireNonNull(journey, "journey");
        if (!journey.isObject() || !journey.path("state").isTextual()) {
            throw MobileIntegrationException.of(MobileIntegrationException.Code.DECODE_FAILURE);
        }
        String state = journey.path("state").textValue();
        return switch (state) {
            case "settled", "completed" -> {
                if (!journey.path("receipt_ref").isTextual()) {
                    throw MobileIntegrationException.of(MobileIntegrationException.Code.DECODE_FAILURE);
                }
                yield settle(journey.path("receipt_ref").textValue(), expectation);
            }
            case "failed", "expired", "refused" -> new Refused(state);
            default -> {
                if (!journey.path("journey_id").isTextual()) {
                    throw MobileIntegrationException.of(MobileIntegrationException.Code.DECODE_FAILURE);
                }
                yield new Pending(journey.path("journey_id").textValue());
            }
        };
    }
}
